File: src/main/java/com/teamwizardry/librarianlib/gui/components/Component3DView.java
package com.teamwizardry.librarianlib.gui.components;

import net.minecraft.client.renderer.GlStateManager;
import net.minecraft.client.renderer.RenderHelper;
import net.minecraft.util.math.Vec3d;

import org.lwjgl.opengl.GL11;

import com.teamwizardry.librarianlib.gui.EnumMouseButton;
import com.teamwizardry.librarianlib.gui.GuiComponent;
import com.teamwizardry.librarianlib.math.Matrix4;
import com.teamwizardry.librarianlib.math.Vec2d;

public class Component3DView extends GuiComponent<Component3DView> {

	public Component3DView(int posX, int posY, int width, int height) {
		super(posX, posY, width, height);
		
		preDraw.add(this::preDraw);
    	postDraw.add(this::postDraw);
	}

    Vec2d dragStart = Vec2d.ZERO;
    public Vec3d offset = Vec3d.ZERO;
    public double zoom;
	public double rotX, rotY, rotZ;
    EnumMouseButton dragButton = null;
	
    @Override
    public void mouseWheel(Vec2d mousePos, int direction) {
    	if(!mouseOverThisFrame)
    		return;
    	if(direction > 0 && zoom < 100) {
    		zoom *= 1.5;
    	}
    	if(direction < 0 && zoom > 1) {
    		zoom /= 1.5;
    	}
    }
    
	@Override
	public void mouseDown(Vec2d mousePos, EnumMouseButton button) {
		super.mouseDown(mousePos, button);

    	if(!mouseOverThisFrame)
    		return;
		if(dragButton != null)
    		return;
    	
    	dragStart = mousePos;
    	dragButton = button;
    }
	
	@Override
	public void mouseUp(Vec2d mousePos, EnumMouseButton button) {
		super.mouseUp(mousePos, button);
		
    	calcDrag(mousePos, button);
    	
    	dragStart = Vec2d.ZERO;
    	dragButton = null;
    }
	
	@Override
	public void drawComponent(Vec2d mousePos, float partialTicks) {
		//noop for now
	}
	
	public void calcDrag(Vec2d mousePos, EnumMouseButton button) {
		if(button != dragButton)
    		return;
    	
    	if(dragButton == EnumMouseButton.LEFT) {
    		rotY += (mousePos.x-dragStart.x);
    		rotX += (mousePos.y-dragStart.y);
    	}
    	if(dragButton == EnumMouseButton.RIGHT) {
    		Vec3d offset = new Vec3d(mousePos.x-dragStart.x, mousePos.y-dragStart.y, 0);
        	Matrix4 matrix = new Matrix4();
        	matrix.rotate(-Math.toRadians(rotZ), new Vec3d(0, 0, 1));
        	matrix.rotate(-Math.toRadians(rotY), new Vec3d(0, 1, 0));
        	matrix.rotate(-Math.toRadians(rotX), new Vec3d(1, 0, 0));
        	matrix.scale(new Vec3d(1.0/zoom, -1.0/zoom, 1.0/zoom));
        	offset = matrix.apply(offset);
        	
        	this.offset = this.offset.add(offset);
    	}
	}
	
	public void preDraw(Component3DView _comp_, Vec2d mousePos, float ticks) {
		GlStateManager.pushMatrix();
        GlStateManager.translate(this.size.x / 2, this.size.y / 2, 500);
        
        double rotX = this.rotX;
        double rotY = this.rotY;
        double rotZ = this.rotZ;
        Vec3d offset = this.offset;
        
        calcDrag(mousePos, dragButton);
        
    	GlStateManager.enableRescaleNormal();
    	GlStateManager.scale(zoom, -zoom, zoom);
    	GlStateManager.rotate((float)this.rotX, 1, 0, 0);
    	GlStateManager.rotate((float)this.rotY, 0, 1, 0);
    	GlStateManager.rotate((float)this.rotZ, 0, 0, 1);
    	GlStateManager.translate(this.offset.xCoord, this.offset.yCoord, this.offset.zCoord);
    	
    	this.offset = offset;
    	this.rotZ = rotZ;
    	this.rotY = rotY;
    	this.rotX = rotX;
    	
    	{ // RenderHelper.enableStandardItemLighting but brighter because of different light and ambiant values.
            Vec3d LIGHT0_POS = new Vec3d(0, 1,  0.1).normalize();//(new Vec3d(0.20000000298023224D, 1.0D, -0.699999988079071D)).normalize();
            Vec3d LIGHT1_POS = new Vec3d(0, 1, -0.1).normalize();//(new Vec3d(-0.20000000298023224D, 1.0D, 0.699999988079071D)).normalize();

            GlStateManager.enableLighting();
            GlStateManager.enableLight(0);
//            GlStateManager.enableLight(1);
            GlStateManager.enableColorMaterial();
            GlStateManager.colorMaterial(GL11.GL_FRONT_AND_BACK, GL11.GL_AMBIENT_AND_DIFFUSE);

            float light = 0.3F;
            float ambiant = 0.7F;
            GlStateManager.glLight(GL11.GL_LIGHT0, GL11.GL_POSITION, RenderHelper.setColorBuffer((float) LIGHT0_POS.xCoord, (float) LIGHT0_POS.yCoord, (float) LIGHT0_POS.zCoord, 0.0f));
            GlStateManager.glLight(GL11.GL_LIGHT0, GL11.GL_DIFFUSE, RenderHelper.setColorBuffer(light, light, light, 1.0F));
            GlStateManager.glLight(GL11.GL_LIGHT0, GL11.GL_AMBIENT, RenderHelper.setColorBuffer(0.0F, 0.0F, 0.0F, 1.0F));
            GlStateManager.glLight(GL11.GL_LIGHT0, GL11.GL_SPECULAR, RenderHelper.setColorBuffer(0.0F, 0.0F, 0.0F, 1.0F));

            GlStateManager.glLight(GL11.GL_LIGHT1, GL11.GL_POSITION, RenderHelper.setColorBuffer((float) LIGHT1_POS.xCoord, (float) LIGHT1_POS.yCoord, (float) LIGHT1_POS.zCoord, 0.0f));
            GlStateManager.glLight(GL11.GL_LIGHT1, GL11.GL_DIFFUSE, RenderHelper.setColorBuffer(light, light, light, 1.0F));
            GlStateManager.glLight(GL11.GL_LIGHT1, GL11.GL_AMBIENT, RenderHelper.setColorBuffer(0.0F, 0.0F, 0.0F, 1.0F));
            GlStateManager.glLight(GL11.GL_LIGHT1, GL11.GL_SPECULAR, RenderHelper.setColorBuffer(0.0F, 0.0F, 0.0F, 1.0F));

            GlStateManager.shadeModel(GL11.GL_FLAT);
            GlStateManager.glLightModel(GL11.GL_LIGHT_MODEL_AMBIENT, RenderHelper.setColorBuffer(ambiant, ambiant, ambiant, 1.0F));
    	}
    	
	}
	
	public void postDraw(Component3DView _comp_, Vec2d mousePos, float ticks) {
		RenderHelper.disableStandardItemLighting();
    	GlStateManager.disableRescaleNormal();
		GlStateManager.popMatrix();
	}

}

File: src/main/java/com/teamwizardry/librarianlib/client/gui/components/Component3DView.kt
package com.teamwizardry.librarianlib.client.gui.components

import com.teamwizardry.librarianlib.client.gui.EnumMouseButton
import com.teamwizardry.librarianlib.client.gui.GuiComponent
import com.teamwizardry.librarianlib.common.util.math.Matrix4
import com.teamwizardry.librarianlib.common.util.math.Vec2d
import net.minecraft.client.renderer.GlStateManager
import net.minecraft.client.renderer.RenderHelper
import net.minecraft.util.math.Vec3d
import org.lwjgl.opengl.GL11

class Component3DView(posX: Int, posY: Int, width: Int, height: Int) : GuiComponent<Component3DView>(posX, posY, width, height) {

    init {
        BUS.hook(PreDrawEvent::class.java) { event -> preDraw(event.mousePos) }
        BUS.hook(PostDrawEvent::class.java) { postDraw() }

        BUS.hook(MouseWheelEvent::class.java) { event ->
            if (mouseOver) {
                if (event.direction == MouseWheelDirection.UP && zoom < 100) {
                    zoom *= 1.5
                }
                if (event.direction == MouseWheelDirection.DOWN && zoom > 1) {
                    zoom /= 1.5
                }
            }
        }

        BUS.hook(MouseUpEvent::class.java) { event ->
            calcDrag(event.mousePos, event.button)

            dragStart = Vec2d.ZERO
            dragButton = null
        }
    }

    internal var dragStart = Vec2d.ZERO
    var offset = Vec3d.ZERO
    var zoom: Double = 0.toDouble()
    var rotX: Double = 0.toDouble()
    var rotY: Double = 0.toDouble()
    var rotZ: Double = 0.toDouble()
    internal var dragButton: EnumMouseButton? = null

    override fun mouseDown(mousePos: Vec2d, button: EnumMouseButton) {
        super.mouseDown(mousePos, button)

        if (!mouseOver)
            return
        if (dragButton != null)
            return

        dragStart = mousePos
        dragButton = button
    }

    override fun drawComponent(mousePos: Vec2d, partialTicks: Float) {
        //NO-OP
    }

    fun calcDrag(mousePos: Vec2d, button: EnumMouseButton?) {
        if (button != dragButton)
            return

        if (dragButton == EnumMouseButton.LEFT) {
            rotY += mousePos.x - dragStart.x
            rotX += mousePos.y - dragStart.y
        }
        if (dragButton == EnumMouseButton.RIGHT) {
            var offset = Vec3d(mousePos.x - dragStart.x, mousePos.y - dragStart.y, 0.0)
            val matrix = Matrix4()
            matrix.rotate(-Math.toRadians(rotZ), Vec3d(0.0, 0.0, 1.0))
            matrix.rotate(-Math.toRadians(rotY), Vec3d(0.0, 1.0, 0.0))
            matrix.rotate(-Math.toRadians(rotX), Vec3d(1.0, 0.0, 0.0))
            matrix.scale(Vec3d(1.0 / zoom, -1.0 / zoom, 1.0 / zoom))
            offset = matrix.apply(offset)

            this.offset = this.offset.add(offset)
        }
    }

    fun preDraw(mousePos: Vec2d) {
        GlStateManager.pushMatrix()
        GlStateManager.translate(this.size.x / 2, this.size.y / 2, 500.0)

        val rotX = this.rotX
        val rotY = this.rotY
        val rotZ = this.rotZ
        val offset = this.offset

        calcDrag(mousePos, dragButton)

        GlStateManager.enableRescaleNormal()
        GlStateManager.scale(zoom, -zoom, zoom)
        GlStateManager.rotate(this.rotX.toFloat(), 1f, 0f, 0f)
        GlStateManager.rotate(this.rotY.toFloat(), 0f, 1f, 0f)
        GlStateManager.rotate(this.rotZ.toFloat(), 0f, 0f, 1f)
        GlStateManager.translate(this.offset.xCoord, this.offset.yCoord, this.offset.zCoord)

        this.offset = offset
        this.rotZ = rotZ
        this.rotY = rotY
        this.rotX = rotX

        run { // RenderHelper.enableStandardItemLighting but brighter because of different light and ambiant values.
            val LIGHT0_POS = Vec3d(0.0, 1.0, 0.1).normalize()//(new Vec3d(0.20000000298023224D, 1.0D, -0.699999988079071D)).normalize();
            val LIGHT1_POS = Vec3d(0.0, 1.0, -0.1).normalize()//(new Vec3d(-0.20000000298023224D, 1.0D, 0.699999988079071D)).normalize();

            GlStateManager.enableLighting()
            GlStateManager.enableLight(0)
            //            GlStateManager.enableLight(1);
            GlStateManager.enableColorMaterial()
            GlStateManager.colorMaterial(GL11.GL_FRONT_AND_BACK, GL11.GL_AMBIENT_AND_DIFFUSE)

            val light = 0.3f
            val ambiant = 0.7f
            GlStateManager.glLight(GL11.GL_LIGHT0, GL11.GL_POSITION, RenderHelper.setColorBuffer(LIGHT0_POS.xCoord.toFloat(), LIGHT0_POS.yCoord.toFloat(), LIGHT0_POS.zCoord.toFloat(), 0.0f))
            GlStateManager.glLight(GL11.GL_LIGHT0, GL11.GL_DIFFUSE, RenderHelper.setColorBuffer(light, light, light, 1.0f))
            GlStateManager.glLight(GL11.GL_LIGHT0, GL11.GL_AMBIENT, RenderHelper.setColorBuffer(0.0f, 0.0f, 0.0f, 1.0f))
            GlStateManager.glLight(GL11.GL_LIGHT0, GL11.GL_SPECULAR, RenderHelper.setColorBuffer(0.0f, 0.0f, 0.0f, 1.0f))

            GlStateManager.glLight(GL11.GL_LIGHT1, GL11.GL_POSITION, RenderHelper.setColorBuffer(LIGHT1_POS.xCoord.toFloat(), LIGHT1_POS.yCoord.toFloat(), LIGHT1_POS.zCoord.toFloat(), 0.0f))
            GlStateManager.glLight(GL11.GL_LIGHT1, GL11.GL_DIFFUSE, RenderHelper.setColorBuffer(light, light, light, 1.0f))
            GlStateManager.glLight(GL11.GL_LIGHT1, GL11.GL_AMBIENT, RenderHelper.setColorBuffer(0.0f, 0.0f, 0.0f, 1.0f))
            GlStateManager.glLight(GL11.GL_LIGHT1, GL11.GL_SPECULAR, RenderHelper.setColorBuffer(0.0f, 0.0f, 0.0f, 1.0f))

            GlStateManager.shadeModel(GL11.GL_FLAT)
            GlStateManager.glLightModel(GL11.GL_LIGHT_MODEL_AMBIENT, RenderHelper.setColorBuffer(ambiant, ambiant, ambiant, 1.0f))
        }

    }

    fun postDraw() {
        RenderHelper.disableStandardItemLighting()
        GlStateManager.disableRescaleNormal()
        GlStateManager.popMatrix()
    }

}
